gnpy/core/network.py

```diff
@@ -222,6 +239,8 @@ def target_power(network, node, equipment): #get_fiber_dp
 
     if isinstance(node, Roadm):
         dp = 0
+    elif isinstance(node, Transceiver):
+        dp = -15
```

Comment:
Current GNPy works for OSNR limited systems, meaning that it is assumed that the power level is guaranteed on receiver. Typical received power is around -15dBm. so we added this condition in this function. We saw this issue when testing the no-amp feature.

Reply:
Nice catch, then. However, right now the fix is buried in an unrelated commit and this comment is the only mention of why it is important. Can you please move this to a new pull request and explain in the commit message what you typed here?

I suspect that different transceivers/transponders have different threshold about a minimal RX power. Is that so, or am I wrong?

Also, it would be nice thing to cover this in an auto test -- that the transponders never end up with something like -20dBm at their ingress port.

Reply:
May we postpone this ?
I would really like to have this PR merged and anyhow working in power limitation on transceiver would need a whole work, which is not a priority right now.
I would prefer to keep as is and add an issue to track this for enhancement and further work.

Reply:
I will create a small pull request doing just this change with a commit message clearly explaining why it is needed. This (#230) pull request can be then merged "on top of the by-then-updated `develop` branch, and the history will be less confusing for a future maintainer. Give me 5 minutes, OK? :)

Reply:
OK !

Reply:
So this is more complex than I anticipated. If I simply add this on top current develop:
```diff
diff --git a/gnpy/core/network.py b/gnpy/core/network.py
index fec0fb0..bae429a 100644
--- a/gnpy/core/network.py
+++ b/gnpy/core/network.py
@@ -222,6 +222,8 @@ def target_power(network, node, equipment): #get_fiber_dp
 
     if isinstance(node, Roadm):
         dp = 0
+    elif isinstance(node, Transceiver):
+        dp = -15
 
     return dp
 
```
...then I'm getting a bunch of failures in `test_amplifier.py` such as this one:
```

gain = 15, nf_expected = 10
setup_edfa_fixed_gain = Edfa(uid='Edfa2', type_variety='test_fixed_gain', interpol_dgt=array([2.71452668, 2.71452668, 2.71452668, 2.71452668, ..., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), pin_db=4.822712330395684, pout_db=5.066598484075778)
si = SpectralInformation(pref=Pref(p_span0=0.0, p_spani=0.0, neq_ch=19.822712330395685), carriers=[Channel(channel_number=1...=45000000000.0, roll_off=0.15, power=Power(signal=0.001, nonlinear_interference=0, amplified_spontaneous_emission=0))])

    @pytest.mark.parametrize("gain, nf_expected", [(15, 10), (20, 5), (25, 5)])
    def test_fixed_gain_nf(gain, nf_expected, setup_edfa_fixed_gain, si):
        """=> unitary test for fixed gain model Edfa._calc_nf() (and Edfa.interpol_params)"""
        edfa = setup_edfa_fixed_gain
        frequencies = array([c.frequency for c in si.carriers])
        pin = array([c.power.signal+c.power.nli+c.power.ase for c in si.carriers])
        pin = pin/db2lin(gain)
        baud_rates = array([c.baud_rate for c in si.carriers])
        edfa.operational.gain_target = gain
        pref = Pref(0, -gain, lin2db(len(frequencies)))
        edfa.interpol_params(frequencies, pin, baud_rates, pref)
    
>       assert pytest.approx(nf_expected, abs=0.01) == edfa.nf[0]
E       assert 10 ± 1.0e-02 == 25.0
E        +  where 10 ± 1.0e-02 = <function approx at 0x7f1be7017510>(10, abs=0.01)
E        +    where <function approx at 0x7f1be7017510> = pytest.approx

tests/test_amplifier.py:99: AssertionError
```
For me, that alone is a good reason to look at this change with even more suspicion than I had before :). So now I think we have two options:
- I just rubber-stamp this change as-is, based on your professional opinion that it's a good change. You clearly know more about this code than I do. At the same time, I do not like this option because I'm essentially merging code I do not understand.
- You help me understand what this thing (-15dBm minimal for transponders) really depends on. Does it depend on the `Fused` node? We separate this into a really smallest possible change, explain in the git log why it is happening, ensure that tests pass, and call it a day. Then we merge this (#230) change.

Reply:
The error in test files is normal: reducing the target power has an big impact on the NF of the last edfa in tests where the trx is directly connected to an edfa. In order to keep performance unchanged and have a more realistic setup, an attenuator has to be put in fromt of the receiver. This is what I just pushed in PR #244 : the extraction from current PR of the changes related to this -15 dB + an attenuator added in the edfa_example_network.json
Hope this is clearer now